libbeat/outputs/shipper/shipper.go

```diff
@@ -70,184 +85,242 @@ func makeShipper(
-	if len(retries) == 0 {
-		batch.ACK()
-		st.Acked(len(nonDroppedEvents))
-		c.log.Debugf("%d events have been accepted, %d dropped", len(nonDroppedEvents), droppedCount)
-	} else {
-		batch.RetryEvents(retries) // decreases TTL unless guaranteed delivery
-		st.Failed(len(retries))
-		c.log.Debugf("%d events have been accepted, %d sent for retry, %d dropped", publishReply.AcceptedCount, len(retries), droppedCount)
-	}
+	s.log.Debugf("total of %d events have been accepted from batch, %d dropped", convertedCount, droppedCount)
+
+	s.pendingMutex.Lock()
+	s.pending = append(s.pending, pendingBatch{
+		batch:        batch,
+		index:        lastAcceptedIndex,
+		serverID:     s.serverID,
+		droppedCount: droppedCount,
+	})
+	s.pendingMutex.Unlock()
 
 	return nil
 }
 
 // Close closes the connection to the shipper server.
 // Also, implements `outputs.Client`
-func (c *shipper) Close() error {
-	if c.client == nil {
+func (s *shipper) Close() error {
+	if s.client == nil {
 		return fmt.Errorf("connection is not established")
 	}
-	err := c.conn.Close()
-	c.conn = nil
-	c.client = nil
+	s.backgroundCancel()
+	err := s.conn.Close()
+	s.conn = nil
+	s.client = nil
 
 	return err
 }
 
 // String implements `outputs.Client`
-func (c *shipper) String() string {
+func (s *shipper) String() string {
 	return "shipper"
 }
 
+func (s *shipper) ackLoop(ctx context.Context, ackClient sc.Producer_PersistedIndexClient) error {
+	st := s.observer
+
+	for {
+		select {
+
+		case <-ctx.Done():
+			return ctx.Err()
+
+		default:
+			// this sends an update and unblocks only if the `PersistedIndex` value has changed
+			indexReply, err := ackClient.Recv()
+			if err != nil {
+				return fmt.Errorf("acknowledgment failed due to the connectivity error: %w", err)
+			}
+
+			s.pendingMutex.Lock()
+			lastProcessed := 0
+			for _, p := range s.pending {
+				if p.serverID != indexReply.Uuid {
+					s.log.Errorf("acknowledgment failed due to a connection to a different server %s, batch was accepted by %s", indexReply.Uuid, p.serverID)
+					p.batch.Cancelled()
+					st.Cancelled(len(p.batch.Events()))
+					lastProcessed++
+					continue
+				}
+
+				// if we met a batch that is ahead of the persisted index
+				// we stop iterating and wait for another update from the server.
+				// The latest pending batch has the max(AcceptedIndex).
+				if p.index > indexReply.PersistedIndex {
+					break
+				}
+
+				p.batch.ACK()
+				ackedCount := len(p.batch.Events()) - p.droppedCount
+				st.Acked(ackedCount)
+				s.log.Debugf("%d events have been acknowledged, %d dropped", ackedCount, p.droppedCount)
+				lastProcessed++
+			}
+			// so we don't perform any manipulation when the pending list is empty
+			// or none of the batches were acknowledged by this persisted index update
+			if lastProcessed != 0 {
+				copy(s.pending[0:], s.pending[lastProcessed:])
+				s.pending = s.pending[lastProcessed:]
+			}
+			s.pendingMutex.Unlock()
```

Comment:
The `Lock` call is outside the `for` loop 

```go
			s.pendingMutex.Lock()
			lastProcessed := 0
			for _, p := range s.pending {
```

but this `Unlock` call is inside it, that doesn't seem right. It looks like the mutex stays locked if we hit this path above:

```go
				if p.index > indexReply.PersistedIndex {
					break
				}
```

Reply:
@cmacknz the loop ends on line 318, the `select` block ends on line 320. The `Unlock` call is outside of the loop.

Reply:
But I agree, it should be a part of the select's case, not out of the select. It was wrong.

Reply:
Thanks, I went through this too quickly.